runtime/doc/vim9.txt

```diff
@@ -1698,6 +1701,156 @@ argument type checking: >
+be inferred, then the type is "any": >
+	var d1 = {}		# type is "dict<any>"
+	var d2 = null_dict	# type is "dict<any>"
+
+Declaring a function, see |vim9-func-declaration|, is particularly unique.
+
+						*null-compare*
+For familiar null compare semantics, do not use null_<type> in a comparison,
+then it is easy to distinguish an empty container from a null container: >
+	vim9script
+	def F(arg: list<string> = null_list)
+	    if arg == null
+	       echo "null"
+	    else
+		echo printf("not null, %sempty", empty(arg) ? '' : 'not ')
+	    endif
+	enddef
+	F()		# output: "null"
+	F(null_list)	# output: "null"
+	F([])		# output: "not null, empty"
+	F([0])		# output: "not null, not empty"
+The above function takes a `list of strings` and reports on it.
+Change the above function signature to accept different types of arguments: >
+	def F(arg: list<any> = null_list)   # any type of list
+	def F(arg: any = null)		    # any type
+<
+In the above example, where the goal is to distinguish a null list from an
+empty list, comparing against `null` instead of `null_list` is the correct
+choice. The basic reason is because "null_list == null" and "[] != null".
+Comparing to null_list fails since "[] == null_list". In the following section
+there are details about comparison results.
+
+					*null-details* *null-anomalies*
+This section describes issues about using null and null_<type>; included below
+are the enumerated results of null comparisons. In some cases, if familiar
+with vim9 null semantics, the programmer may chose to use null_<type> in
+comparisons and/or other situations.
+
+Elsewhere in the documentation it says:
+	Quite often a null value is handled the same as an
+	empty value, but not always
+Here's an example: >
+	vim9script
+	var s1: list<string>
+	var s2: list<string> = null_list
+	echo s1		    # output: "[]"
+	echo s2		    # output: "[]"
+	
+	echo s1 + ['a']     # output: "['a']"
+	echo s2 + ['a']     # output: "['a']"
+	
+	echo s1->add('a')   # output: "['a']"
+	echo s2->add('a')   # E1130: Can not add to null list
+<
+Two values equal to a null_<type> are not necessarily equal to each other: >
+	vim9script
+	echo {} == null_dict      # true
+	echo null_dict == null    # true
+	echo {} == null           # false
+<
+Unlike the other containers, an uninitialized string is equal to null. The
+'is' operator can be used to determine if it is a null_string: >
+	vim9script
+	var s1: string
+	var s2 = null_string
+	echo s1 == null		# true - this is unexpected
+	echo s2 == null		# true
+	echo s2 is null_string	# true
+
+	var b1: blob
+	var b2 = null_blob
+	echo b1 == null		# false
+	echo b2 == null		# true
+<
+Any variable initialized to the null_<type> is equal to the null_<type> and is
+also equal to null. For example: >
+	vim9script
+	var x = null_blob
+	echo x == null_blob	# true
+	echo x == null		# true
+<
+An uninitialized variable is usually equal to null; it depends on its type:
+	var s: string		s == null
```

Comment:
this maybe was a bug.

Reply:
Maybe, open an issue if you want. This is how it works in vim9.0.

Reply:
there no v9.0, v9.0 never published, only v8.2 and the coming v9.1
---
anyway, @yegappan the instr from bram just say non-init string should be 'empty', but emtpy did not mean 'null', so '' (emtpy string) should be more make sense?

Reply:
When a string variable is not initialized, memory is not allocated for the string.  Internally in the code, the string pointer in typval is NULL (https://github.com/vim/vim/blob/master/src/structs.h#L1629).  So the string is `null` and is not `empty`.

Reply:
> ## there no v9.0, v9.0 never published, only v8.2 and the coming v9.1

What do you mean by v9.0 was never published? See https://www.vim.org/vim90.php

Reply:
> So the string is `null` and is not `empty`.

the 'empty' is not my word, it is bram's instr in the ':h e1134'.
i may feel you and errael thought it too c/c++ style, it is a scripting language, in script world, there is no 'pointer' and deconstructer, and if no init, usually to string it is a "".
// anyway, now i am not against it since i felt it is already difficult to change it, so i have opened PR #13769 to refine the doc a bit, that's all.

Reply:
> > ## there no v9.0, v9.0 never published, only v8.2 and the coming v9.1
> 
> What do you mean by v9.0 was never published? See https://www.vim.org/vim90.php

most recent linux distribution did not use v9.0 yet, it is still v8.2, vs v9.0 mostly was our dev ver.

Reply:
Recent Linux distributions do use Vim 9.0.  Refer to the following page for the Ubuntu Vim package:

https://packages.ubuntu.com/search?keywords=vim

As you can see, starting with the lunar release, Ubuntu started using Vim 9.0.

Similarly, for the Debian Vim package, refer to https://packages.debian.org/search?keywords=vim.
Starting with the bookworm release, Vim 9.0 is used.

Fedora Linux also started using Vim9.0 from the Fedora 37 release.
See https://rpmfind.net/linux/rpm2html/search.php?query=vim-common.

MacOS ships with Vim 9.0.

Vim 9.0 is not a dev-only version.  It ships with all the major distributions.

Reply:
ok, fine, or whatever, i had said many times also worried it was too late to change also.
so anyway, see my previous comment https://github.com/vim/vim/pull/13750#discussion_r1436115577